libevmasm/Inliner.cpp

```diff
@@ -0,0 +1,182 @@
+{
+	std::map<u256, ranges::span<AssemblyItem const>> inlinableBlockItems;
+	std::map<u256, uint64_t> numPushTags;
+	std::optional<size_t> lastTag;
+	for (auto&& [index, item]: _items | ranges::views::enumerate)
+	{
+		// The number of PushTag's approximates the number of calls to a block.
+		if (item.type() == PushTag)
+			numPushTags[item.data()]++;
+
+		// We can only inline blocks with straight control flow that end in a jump.
+		if (lastTag && SemanticInformation::breaksCSEAnalysisBlock(item, true))
+		{
+			if (item == Instruction::JUMP)
+				inlinableBlockItems.emplace(
+					piecewise_construct,
+					forward_as_tuple(_items[*lastTag].data()),
+					forward_as_tuple(_items | ranges::views::slice(*lastTag + 1, index + 1))
+				);
+			lastTag.reset();
+		}
+
+		if (item.type() == Tag)
+			lastTag = index;
+	}
+
+	// Filter candidates for general inlinability and store the number of PushTag's alongside the assembly items.
+	map<u256, InlinableBlock> result;
+	for (auto&& [tag, items]: inlinableBlockItems)
+		if (uint64_t const* numPushes = util::valueOrNullptr(numPushTags, tag))
+		{
+			InlinableBlock block{items, *numPushes};
+			if (isInlineCandidate(tag, block))
+				result.emplace(std::make_pair(tag, InlinableBlock{items, *numPushes}));
+		}
+	return result;
+}
+
+namespace
+{
+optional<AssemblyItem::JumpType> determineJumpType(AssemblyItem::JumpType _intoBlock, AssemblyItem::JumpType _outOfBlock)
+{
+	auto jumpTypeToInt = [](AssemblyItem::JumpType _jumpType) -> int {
+		switch (_jumpType)
+		{
+			case AssemblyItem::JumpType::IntoFunction:
+				return +1;
+			case AssemblyItem::JumpType::OutOfFunction:
+				return -1;
+			case AssemblyItem::JumpType::Ordinary:
+				return 0;
+		}
+		assertThrow(false, OptimizerException, "");
+	};
+	switch (jumpTypeToInt(_intoBlock) + jumpTypeToInt(_outOfBlock))
+	{
+		case 0:
+			return AssemblyItem::JumpType::Ordinary;
+		case 1:
+			return AssemblyItem::JumpType::IntoFunction;
+		case -1:
+			return AssemblyItem::JumpType::OutOfFunction;
+		default:
+			return nullopt;
+	}
+}
+}
+
+optional<AssemblyItem> Inliner::shouldInline(u256 const&, AssemblyItem const& _jump, InlinableBlock const& _block) const
+{
+	// Determine the exit jump to be used, if the block is inlined.
+	AssemblyItem exitJump = _block.items.back();
+	if (auto exitJumpType = determineJumpType(_jump.getJumpType(), exitJump.getJumpType()))
+		exitJump.setJumpType(*exitJumpType);
+	else
+		return nullopt;
+
+	// Always try to inline if there is at most one call to the block.
+	if (_block.pushTagCount == 1)
+		return exitJump;
+
+	// Always inline small blocks.
+	if (static_cast<size_t>(_block.items.size()) <= m_inlineMaxOpcodes)
+		return exitJump;
+
+	return nullopt;
+}
+
+
+void Inliner::optimise()
+{
+	std::map<u256, InlinableBlock> inlinableBlocks = determineInlinableBlocks(m_items);
+
+	if (inlinableBlocks.empty())
+		return;
+
+	AssemblyItems newItems;
+	for (auto it = m_items.begin(); it != m_items.end(); ++it)
```

Comment:
Actualyl can't we use `iterateReplacing` here?

Reply:
``iterateReplacingWindow`` modifies the ``vector`` we iterate over in-place - so that would invalidate the ``ranges::span``s into ``m_items`` here... so yes, we could use it, but then we would need to copy the items into ``InlinableBlock`` instead of only storing ``span``s into ``m_items`` there. Both fine with me...

Reply:
Ah right, then it is probably not a good idea.